funcx_endpoint/funcx_endpoint/endpoint/interchange.py

```diff
@@ -351,49 +351,79 @@ def _main_loop(self):
-                    "to forwarder queues"
-                )
-                continue
+
+        num_tasks_forwarded = 0
+        num_results_forwarded = 0
+
+        def process_pending_tasks():
+            nonlocal num_tasks_forwarded
+            while not (self._quiesce_event.is_set() or self._kill_event.is_set()):
+                try:
+                    task = self.pending_task_queue.get(timeout=1)
+                    executor.submit_raw(task)
+                    num_tasks_forwarded += 1  # Safe given GIL
+
+                except queue.Empty:
+                    continue
+
+                except Exception:
+                    log.exception("Unhandled issue while waiting for pending tasks")
+            log.debug("Exit pending task processor thread.")
+
+        def process_pending_results():
+            nonlocal num_results_forwarded
+            while not (self._quiesce_event.is_set() or self._kill_event.is_set()):
+                try:
+                    result = self.results_passthrough.get(timeout=1)
+
+                    task_id = result["task_id"]
+                    if task_id:
+                        self.results_ack_handler.put(task_id, results["message"])
+                        log.info(f"Passing result to forwarder for task {task_id}")
+
+                    # results will be a pickled dict with task_id, container_id,
+                    # and results/exception
+                    self.results_outgoing.publish(results["message"])
+                    num_results_forwarded += 1  # Safe given GIL
+
+                except queue.Empty:
+                    continue
+
+                except Exception:
+                    log.exception(
+                        "Something broke while forwarding results from executor "
+                        "to forwarder queues"
+                    )
+                    continue
+            log.debug("Exit pending result processor thread.")
+
+        task_processor_thread = threading.Thread(
+            target=process_pending_tasks, name="Pending Task Handler"
+        )
+        result_processor_thread = threading.Thread(
+            target=process_pending_results, name="Pending Result Handler"
+        )
+        task_processor_thread.start()
+        result_processor_thread.start()
+
+        last_t, last_r = 0, 0
+        while not self._quiesce_event.wait(self.heartbeat_threshold):
+            # Possibly TOCTOU here, but we don't need to be super precise.  The
+            # point here is to mention "still alive" and that we're still working
+            num_t, num_r = num_tasks_forwarded, num_results_forwarded
+            log.debug(
+                "Heartbeat.  Approximate Tasks and Results forwarded since last "
+                "heartbeat: %s (T), %s (R)",
+                num_t - last_t,
+                num_r - last_r,
+            )
+            last_t, last_r = num_t, num_r
+
+        task_processor_thread.join()
+        result_processor_thread.join()
```

Comment:
`join()` with no args could block indefinitely if the thread is broken -- e.g. we have a logical error which infinite loops.

I think we should `join(timeout=...)` and check `is_alive()` to ensure that if these do not exit, we terminate the process (which will forcibly kill any threads) and log an error message.

Reply:
I've added this, but at this point, it's moot: the threads are watching the same event as the previous loop.  These joins will not be attempted until the event has been set.